### PR TITLE
feat: Code action for toggling advisory configuration diagnostics

### PR DESCRIPTION
**Pull request by jirimosinger**

### CODE REVIEW

File: language_server/src/main.cpp [hunk 1]
```diff
@@ -82,10 +84,12 @@ class main_program final : public json_sink,
         , router(&lsp_queue)
         , dap_sessions(*this, json_output, &dap_telemetry_broker)
         , virtual_files(*ws_mngr, json_output)
+        , config_diags_provider(*ws_mngr)
     {
         router.register_route(dap_sessions.get_filtering_predicate(), dap_sessions);
         router.register_route(virtual_files.get_filtering_predicate(), virtual_files);
         router.register_route(external_files.get_filtering_predicate(), external_files);
+        router.register_route(config_diags_provider.get_filtering_predicate(), config_diags_provider);
```
Comment (slavek-kucera) [inline language_server/src/main.cpp:92]: Is there a particular reason for going this router? The processing in WSmgr is now fully interruptible and it is possible to "jump" the queue there or execute synchronously without queueing.

Comment (jirimosinger) [reply language_server/src/main.cpp:92]: No, no particular reason. I'll take a look.

File: clients/vscode-hlasmplugin/src/extension.ts [hunk 2]
```diff
@@ -186,7 +187,7 @@ async function registerToContext(context: vscode.ExtensionContext, client: vscod
     // register provider for all hlasm debug configurations
     context.subscriptions.push(vscode.debug.registerDebugConfigurationProvider('hlasm', new HLASMConfigurationProvider()));
     context.subscriptions.push(vscode.debug.registerDebugAdapterDescriptorFactory('hlasm', new HLASMDebugAdapterFactory(client)));
-    context.subscriptions.push(vscode.languages.registerCodeActionsProvider('hlasm', new HLASMCodeActionsProvider(client)));
+    context.subscriptions.push(vscode.languages.registerCodeActionsProvider(['hlasm', 'json'], new HLASMCodeActionsProvider(client)));
```
Comment (slavek-kucera) [inline clients/vscode-hlasmplugin/src/extension.ts:190]: I wonder whether the document selector should be more specific to only include our JSONs....

File: clients/vscode-hlasmplugin/src/hlasmConfigurationDiagnosticsProvider.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+
+import * as vscodelc from 'vscode-languageclient';
+
+export function toggleNonCriticalConfigurationDiagnostics(client: vscodelc.BaseLanguageClient) {
+    client.sendRequest<void>("toggle_non_critical_configuration_diagnostics");
```
Comment (slavek-kucera) [inline clients/vscode-hlasmplugin/src/hlasmConfigurationDiagnosticsProvider.ts:18]: I think this should be sendNotification - we are not expecting answer and the server does not provide one.

File: parser_library/src/workspaces/workspace_configuration.cpp [hunk 4]
```diff
@@ -853,18 +926,42 @@ const program* workspace_configuration::get_program(const utils::resource::resou
...
     {
         if (std::regex_match(file_location_normalized.get_uri(), pattern))
-            return { &program.pgm, false };
+            return try_retrieve_program(pgm_details, false);
     }
+
     return { nullptr, false };
 }
 
+const workspace_configuration::missing_pgroup_details* workspace_configuration::get_missing_pgroup_details_normalized(
```
Comment (slavek-kucera) [inline parser_library/src/workspaces/workspace_configuration.cpp:950]: As there does not seem to be a version that does not take the non-normalize file_location, I'd just drop the normalized suffix.

File: parser_library/src/workspaces/workspace_configuration.cpp [hunk 5]
```diff
@@ -943,17 +1040,21 @@ void workspace_configuration::update_external_configuration(
         });
 }
 
-
 void workspace_configuration::prune_external_processor_groups(const utils::resource::resource_location& location)
 {
+    constexpr auto is_erase_candidate = [](const program_related_details& pgm_rel_details) {
+        const auto& tagged_pgm = std::get_if<tagged_program>(&pgm_rel_details);
+        return tagged_pgm && tagged_pgm->pgm.external;
+    };
```
Comment (slavek-kucera) [inline parser_library/src/workspaces/workspace_configuration.cpp:1048]:
```suggestion
    constexpr auto is_external = [](const program_related_details& pgm_rel_details) {
        const auto* tagged_pgm = std::get_if<tagged_program>(&pgm_rel_details);
        return tagged_pgm && tagged_pgm->pgm.external;
    };
```

File: parser_library/test/workspace/b4g_integration_test.cpp [hunk 6]
```diff
@@ -459,3 +471,112 @@ TEST(b4g_integration_test, b4g_conf_noproc_proc_group_default)
...
+
+    run_if_valid(ws.did_open_file(pgm_a));
+    parse_all_files(ws);
+    ws.collect_diags();
+    EXPECT_TRUE(ws.diags().empty());
+
+    ws.include_non_critical_configuration_diagnostics(true);
+    ws.diags().clear();
+    ws.collect_diags();
+    EXPECT_TRUE(matches_message_codes(ws.diags(), { "B4G003" }));
```
Comment (slavek-kucera) [inline parser_library/test/workspace/b4g_integration_test.cpp:576]: This is for the `P2` default group, right?

File: clients/vscode-hlasmplugin/src/constants.ts [hunk 7]
```diff
@@ -12,8 +12,14 @@
...
+import * as vscode from 'vscode';
+
 export const pgm_conf_file = 'pgm_conf.json';
 export const proc_grps_file = 'proc_grps.json';
 export const bridge_json_file = '.bridge.json';
 export const hlasmplugin_folder = '.hlasmplugin';
 export const ebg_folder = '.ebg';
+
+export const pgm_conf_filter: vscode.DocumentFilter = { language: 'json', pattern: '**/pgm_conf.json' };
+export const proc_grps_filter: vscode.DocumentFilter = { language: 'json', pattern: '**/proc_grps.json' };
```
Comment (slavek-kucera) [inline clients/vscode-hlasmplugin/src/constants.ts:24]: Does it work with `**/.hlasmplugin/....`?

File: parser_library/test/workspace/b4g_integration_test.cpp (hunk 6 above)
Comment (jirimosinger) [reply parser_library/test/workspace/b4g_integration_test.cpp:576]: Yup, that's right

Comment (slavek-kucera) [reply parser_library/test/workspace/b4g_integration_test.cpp:576]: That's ok... I just wanted to make sure I understood the test.

File: clients/vscode-hlasmplugin/src/hlasmCodeActionsProvider.ts [hunk 8]
```diff
@@ -25,6 +26,12 @@ export class HLASMCodeActionsProvider implements vscode.CodeActionProvider {
     async provideCodeActions(document: vscode.TextDocument, range: vscode.Range | vscode.Selection, context: vscode.CodeActionContext, token: vscode.CancellationToken): Promise<(vscode.CodeAction | vscode.Command)[]> {
         const result: vscode.CodeAction[] = [];
 
+        const isConfig = document.fileName.endsWith('\\.bridge.json') || document.fileName.endsWith('\\pgm_conf.json');
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply clients/vscode-hlasmplugin/src/hlasmCodeActionsProvider.ts:29]: It's weird to me that it didn't fail during CI tests. Nevertheless, changed to the URI path.

File: clients/vscode-hlasmplugin/src/test/suite/codeActions.test.ts [hunk 9]
```diff
@@ -79,4 +81,53 @@ suite('Code actions', () => {
...
+        const configRelPath = path.relative(helper.getWorkspacePath(), configFileUri.fsPath);
+        let diagnostic_event = helper.waitForDiagnostics(configRelPath, false, diagSource);
+
+        await helper.showDocument(file, 'hlasm');
+
+        helper.assertMatchingMessageCodes(await diagnostic_event, criticalDiags, diagSource);
+
+        let codeActionsList = await queryCodeActions(configFileUri, new vscode.Range(0, 0, 0, 0), 500);
+
+        assert.equal(codeActionsList.length, 1);
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply clients/vscode-hlasmplugin/src/test/suite/codeActions.test.ts:95]: Good point - changed for the new code action tests.

However, it's a bit difficult to fix it for the old ones not all of them have command parameter. The structure looks like this:
image

File: clients/vscode-hlasmplugin/src/test/suite/testHelper.ts [hunk 10]
```diff
@@ -157,7 +157,7 @@ export function timeout(ms: number, error_message: string | undefined = undefine
     return new Promise<void>((_, reject) => { setTimeout(() => reject(error_message && Error(error_message)), ms); });
 }
 
-export async function waitForDiagnostics(file: string | vscode.Uri, nonEmptyOnly: boolean = false) {
+export async function waitForDiagnostics(file: string | vscode.Uri, nonEmptyOnly: boolean = false, source: string | undefined = undefined) {
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply clients/vscode-hlasmplugin/src/test/suite/testHelper.ts:160]: Added the `@deprecated` annotation so it's clearer in VSCode:
image

File: language_server/test/lsp/lsp_server_test.cpp [hunk 11]
```diff
@@ -350,3 +350,14 @@ TEST(lsp_server_test, external_configuration_request_error)
...
+
+TEST(lsp_server_test, toggle_non_critical_configuration_diagnostics)
+{
+    NiceMock<test::ws_mngr_mock> ws_mngr;
+    lsp::server s(ws_mngr);
+
+    EXPECT_CALL(ws_mngr, toggle_non_critical_configuration_diagnostics());
+
+    s.message_received(
+        R"({"jsonrpc":"2.0","method":"toggle_non_critical_configuration_diagnostics","params":{"uri":"scheme:path"}})"_json);
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply language_server/test/lsp/lsp_server_test.cpp:362]: Changed it to the real thing, thx!

File: parser_library/src/workspaces/workspace_configuration.h [hunk 12]
```diff
@@ -103,6 +103,16 @@ struct program
     bool external;
 };
 
+struct configuration_diagnostics_parameters
+{
+    std::unordered_map<utils::resource::resource_location,
+        std::unordered_set<utils::resource::resource_location, utils::resource::resource_location_hasher>,
+        utils::resource::resource_location_hasher>
+        used_configs_opened_files_map;
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/workspaces/workspace_configuration.h:111]: Not sure if I've completely got what you've meant here. 

Anyway, I think that it needs to remain an `unordered_map` for easier config file location lookup, however, I switched the `set` for `vector`

Just to visualize, the structure is as follows:
```
.bridge.json        -> PGM_1
                    -> PGM_2
                    -> PGM_3

pgm_conf.json -> A
              -> B
              -> C
```

**Comment by jirimosinger** [comment]
I'm also thinking about issuing a "No more configuration advisories" info diagnostics so the code action provides at least some feedback to the user when there are no more diagnostics issues. It doesn't need to be in this PR, though. 

What do you think?

File: clients/vscode-hlasmplugin/src/test/suite/testHelper.ts [hunk 13]
```diff
@@ -193,9 +198,14 @@ export async function waitForDiagnosticsChange(file: string | vscode.Uri, action
...
                 return;
-            const diags = vscode.languages.getDiagnostics(forFile);
+            let diags = vscode.languages.getDiagnostics(forFile);
             if (diags.length === initialDiags.length && diags.map(x => JSON.stringify(x)).sort().every((x, i) => x === initialDiags[i]))
                 return;
+            if (source) {
+                diags = diags.filter(d => { return d.source === source });
+                if (diags.length === 0)
+                    return;
+            }
```
Comment (slavek-kucera) [inline clients/vscode-hlasmplugin/src/test/suite/testHelper.ts:208]:
```suggestion
            const diags = vscode.languages.getDiagnostics(forFile);
            if (diags.length === initialDiags.length && diags.map(x => JSON.stringify(x)).sort().every((x, i) => x === initialDiags[i]))
                return;
            if (source && diags.every(d => d.source !== source))
                    return;
```

however I think that if the source is specified all the other code should probably use it as well? The initialialDiags should be filtered, the count test in the callback...

**Comment by slavek-kucera** [comment]
Regarding "No more configuration advisories" - I am not sure about a diagnostic about there not being other diagnostics...

File: clients/vscode-hlasmplugin/src/test/suite/codeActions.test.ts [hunk 14]
```diff
@@ -79,4 +81,53 @@ suite('Code actions', () => {
...
+            .then(codeActionList => codeActionList.filter(x => x.command?.command === 'extension.hlasm-plugin.toggleAdvisoryConfigurationDiagnostics'));
+
+        assert.equal(codeActionsList.length, 1);
+        assert.strictEqual(codeActionsList[0].command!.title, 'Show all configuration diagnostics');
+
+        diags = await helper.waitForDiagnosticsChange(configRelPath, () => { vscode.commands.executeCommand(codeActionsList[0].command!.command, configFileUri, new vscode.Range(0, 0, 0, 0)); }, diagSource);
+
+        helper.assertMatchingMessageCodes(diags, allDiags, diagSource);
+
+        codeActionsList = await queryCodeActions(configFileUri, new vscode.Range(0, 0, 0, 0), 500).then(codeActionList => {
```
Comment (slavek-kucera) [inline clients/vscode-hlasmplugin/src/test/suite/codeActions.test.ts:101]: same thing as the call above_

**Review by slavek-kucera — APPROVED** [review]

File: clients/vscode-hlasmplugin/src/test/suite/codeActions.test.ts (hunk 14 above)
Comment (jirimosinger) [reply clients/vscode-hlasmplugin/src/test/suite/codeActions.test.ts:101]: 🤦‍♂️

**Review by slavek-kucera — APPROVED** [review]

**Pull request merged**